File: artifacts/discord-bot/bot/services/guildService.js
const db = require('../database/sqlite');

/**
 * Default guild configuration values.
 */
const defaultGuild = {
  modules: {
    leveling: true,
    invites: true,
    welcome: true,
    moderation: true,
    afk: true,
    logging: true,
    autorole: true,
    autovc: true,
    ai: true,
  },
  channels: {
    welcome: null,
    leave: null,
    logs: null,
    modLogs: null,
    inviteLogs: null,
    levelLogs: null,
  },
  roles: {
    autorole: null,
    muteRole: null,
    levelRoles: [],
  },
  settings: {
    welcome: {
      message: 'Welcome {user} to **{server}**! You are member #{memberCount}.',
      embed: true,
      dm: false,
      delay: 0,
    },
    leave: {
      message: '**{user}** has left **{server}**. We now have {memberCount} members.',
      embed: false,
    },
    leveling: {
      xpMin: 15,
      xpMax: 25,
      cooldown: 60,
      levelUpMessage: 'Congratulations {user}! You reached level **{level}**!',
      blacklistedChannels: [],
    },
    automod: {
      antiSpam: true,
      antiDuplicate: true,
      antiLinks: false,
      antiCaps: false,
      badWords: [],
      mentionLimit: 5,
      punishment: 'warn',
    },
    invites: { rewards: [] },
    autovc: { category: null, lobby: null },
    ai: {
      chatChannel: null,
      hubChannel: null,
      silenceHours: 6,
      emotionEnabled: true,
    },
  },
};

/**
 * Utility to parse JSON safely
 */
function safeParse(str, def) {
  if (!str) return def;
  try {
    return JSON.parse(str);
  } catch {
    return def;
  }
}

/**
 * Get guild config, creating it with defaults if it doesn't exist.
 */
async function getGuild(guildId) {
  const row = db.prepare('SELECT * FROM guilds WHERE guild_id = ?').get(guildId);

  if (!row) {
    return createGuild(guildId);
  }

  return {
    guild_id: row.guild_id,
    modules: safeParse(row.modules, defaultGuild.modules),
    channels: safeParse(row.channels, defaultGuild.channels),
    roles: safeParse(row.roles, defaultGuild.roles),
    settings: safeParse(row.settings, defaultGuild.settings),
  };
}

/**
 * Create a new guild with default settings.
 */
async function createGuild(guildId) {
  db.prepare('INSERT INTO guilds (guild_id, modules, channels, roles, settings) VALUES (?, ?, ?, ?, ?)').run(
    guildId,
    JSON.stringify(defaultGuild.modules),
    JSON.stringify(defaultGuild.channels),
    JSON.stringify(defaultGuild.roles),
    JSON.stringify(defaultGuild.settings)
  );

  return {
    guild_id: guildId,
    ...defaultGuild,
  };
}

/**
 * Update a specific top-level field (modules, channels, roles, settings) by merging.
 */
async function updateGuildField(guildId, field, updates) {
  const guild = await getGuild(guildId);
  const merged = { ...guild[field], ...updates };

  db.prepare(`UPDATE guilds SET ${field} = ? WHERE guild_id = ?`).run(
    JSON.stringify(merged),
    guildId
  );

  return { ...guild, [field]: merged };
}

/**
 * Update a nested setting inside the settings JSONB column.
 * e.g. updateGuildSetting(guildId, 'leveling', { xpMin: 10 })
 */
async function updateGuildSetting(guildId, key, updates) {
  const guild = await getGuild(guildId);
  const merged = { ...guild.settings, [key]: { ...guild.settings[key], ...updates } };

  db.prepare('UPDATE guilds SET settings = ? WHERE guild_id = ?').run(
    JSON.stringify(merged),
    guildId
  );

  return { ...guild, settings: merged };
}

/**
 * Toggle a module on/off. Returns the new boolean value.
 */
async function toggleModule(guildId, moduleName) {
  const guild = await getGuild(guildId);
  const current = guild.modules[moduleName];
  if (current === undefined) throw new Error(`Unknown module: ${moduleName}`);

  const updated = { ...guild.modules, [moduleName]: !current };

  db.prepare('UPDATE guilds SET modules = ? WHERE guild_id = ?').run(
    JSON.stringify(updated),
    guildId
  );

  return !current;
}

/**
 * Check if a module is enabled for a guild.
 */
async function isModuleEnabled(guildId, moduleName) {
  const guild = await getGuild(guildId);
  return guild.modules[moduleName] === true;
}

module.exports = {
  getGuild,
  createGuild,
  updateGuildField,
  updateGuildSetting,
  toggleModule,
  isModuleEnabled,
};

File: services/guild_service.py
import json
from database.connection import get_db

DEFAULT_MODULES = {
    "moderation": True,
    "leveling": True,
    "tickets": True,
    "giveaways": True,
    "invites": True,
    "reaction_roles": True,
    "button_roles": True,
    "custom_commands": True,
    "welcome": True,
    "auto_vc": True,
    "logging": True
}

DEFAULT_SETTINGS = {
    "automod": {
        "anti_spam": False,
        "anti_links": False,
        "caps_limit": 0,
        "mention_limit": 0
    },
    "leveling": {
        "xp_multiplier": 1.0,
        "min_xp": 15,
        "max_xp": 25,
        "cooldown": 60,
        "rewards": {}
    }
}

async def get_guild(guild_id: str | int) -> dict:
    gid = str(guild_id)
    async with get_db() as db:
        async with db.execute("SELECT * FROM guilds WHERE guild_id = ?", (gid,)) as cursor:
            row = await cursor.fetchone()
            if row:
                return {
                    "guild_id": row["guild_id"],
                    "modules": json.loads(row["modules"] or "{}"),
                    "channels": json.loads(row["channels"] or "{}"),
                    "roles": json.loads(row["roles"] or "{}"),
                    "settings": json.loads(row["settings"] or "{}"),
                }
            
            # Initialize default
            modules_json = json.dumps(DEFAULT_MODULES)
            channels_json = json.dumps({})
            roles_json = json.dumps({})
            settings_json = json.dumps(DEFAULT_SETTINGS)
            
            await db.execute(
                "INSERT OR IGNORE INTO guilds (guild_id, modules, channels, roles, settings) VALUES (?, ?, ?, ?, ?)",
                (gid, modules_json, channels_json, roles_json, settings_json)
            )
            await db.commit()
            return {
                "guild_id": gid,
                "modules": DEFAULT_MODULES.copy(),
                "channels": {},
                "roles": {},
                "settings": DEFAULT_SETTINGS.copy()
            }

async def update_guild(guild_id: str | int, field: str, value: dict):
    gid = str(guild_id)
    # Ensure guild exists
    await get_guild(gid)
    async with get_db() as db:
        await db.execute(
            f"UPDATE guilds SET {field} = ? WHERE guild_id = ?",
            (json.dumps(value), gid)
        )
        await db.commit()

async def is_module_enabled(guild_id: str | int, module_name: str) -> bool:
    guild_data = await get_guild(guild_id)
    return guild_data.get("modules", {}).get(module_name, True)
